telemetry/admin/src/main/scala/io/buoyant/telemetry/admin/AdminTelemeter.scala

```diff
@@ -0,0 +1,121 @@
+
+  val adminHandlers: Admin.Handlers = Seq(
+    "/admin/metrics.json" -> this
+  )
+
+  val stats = NullStatsReceiver
+  def tracer = NullTracer
+
+  private[this] val started = new AtomicBoolean(false)
+  // only run at most once
+  def run(): Closable with Awaitable[Unit] =
+    if (started.compareAndSet(false, true)) run0()
+    else Telemeter.nopRun
+
+  private[this] def run0() = {
+    val task = timer.schedule(snapshotInterval) {
+      summarySnapshots = snapshotHistograms(metrics)
+    }
+
+    new Closable with CloseAwaitably {
+      override def close(deadline: Time): Future[Unit] = closeAwaitably(task.close(deadline))
+    }
+  }
+
+  def apply(request: Request): Future[Response] = {
+    val pretty = request.getBooleanParam("pretty", false)
+    val response = request.response
+    response.mediaType = MediaType.Json
+    response.withOutputStream(writeJson(_, pretty))
+    Future.value(response)
+  }
+
+  private[this] var summarySnapshots = Map.empty[Stat, HistogramSummary]
+
+  private[this] val json = new JsonFactory()
+  private[this] def writeJson(out: OutputStream, pretty: Boolean = false): Unit = {
+    val jg = json.createGenerator(out)
+    if (pretty) jg.setPrettyPrinter(new DefaultPrettyPrinter())
+    jg.writeStartObject()
+    val flattened = if (pretty)
+      flattenMetricsTree(metrics).sortBy(_._1)
+    else
+      flattenMetricsTree(metrics)
+    flattened.foreach {
+      case (name, metric) =>
+        metric match {
+          case c: Counter =>
+            jg.writeNumberField(name, c.get)
+          case g: Gauge =>
+            jg.writeNumberField(name, g.get)
+          case s: Stat =>
+            for (summary <- summarySnapshots.get(s)) {
+              jg.writeNumberField(s"$name.count", summary.count)
+              if (summary.count > 0) {
+                jg.writeNumberField(s"$name.max", summary.max)
+                jg.writeNumberField(s"$name.min", summary.min)
+                jg.writeNumberField(s"$name.p50", summary.p50)
+                jg.writeNumberField(s"$name.p90", summary.p90)
+                jg.writeNumberField(s"$name.p95", summary.p95)
+                jg.writeNumberField(s"$name.p99", summary.p99)
+                jg.writeNumberField(s"$name.p9990", summary.p9990)
+                jg.writeNumberField(s"$name.p9999", summary.p9999)
+                jg.writeNumberField(s"$name.sum", summary.sum)
+                jg.writeNumberField(s"$name.avg", summary.avg)
+              }
+            }
+          case Metric.None =>
+        }
+    }
+    jg.writeEndObject()
+    jg.close()
+  }
+
+  private[this] def flattenMetricsTree(tree: MetricsTree, prefix: String = ""): Seq[(String, Metric)] = {
+    tree.children.foldLeft(Seq(prefix -> tree.metric)) {
+      case (acc, (name, child)) =>
+        if (prefix == "")
+          acc ++ flattenMetricsTree(child, name)
+        else
+          acc ++ flattenMetricsTree(child, s"$prefix/$name")
+    }
+  }
```

Comment:
tioli: I would probably opt to do this procedurally with a `mutable.Buffer[(String, Metric)]` to reduce garbage generated from hitting metrics.json (these things end up being _megabytes_ in production).